### PR TITLE
Motors Update

### PR DESCRIPTION
**Pull request by Baldstrom**
# New Motors Update
There are a lot of changes in this branch, some of which are breaking, but we believe that these changes make using motors and servos in Scarlet more intuitive.

## Major Changes
- Makes motors more user-friendly. For example, TalonMC and Cytron MCs handle input filters and threading internally. 
- `TalonMC` and `CytronMD30C` now take advantage of new `IFilter` requirement: `IsSteadyState()` to handle threading internally.
- `IMotor` and `IServo` now use `SetEnabled(bool Enabled)` to set whether or not the motor runs
- Made filter changes to add `IsSteadyState()` functionality to `LowPass`, `Average`, and `Kalman` filters. 
- Added Testing files for `TalonMC`
- Added `Constants` class to `Utils` for across-code ease-of-use.
- Removed file search utility in `Utilities`

## Notes

- All `TalonMC` tests pass
- Added Testing playlists
- Added `LowPass` test class (not populated)
- Added `TestablePWMOutput` for testing purposes (to see changes directly on the duty cycle output)
- Did not directly test `CytronMD30C`, but it works in the exact same way `TalonMC` does, except that it uses a GPIO pin to set the direction of the motor which was tested when it was originally implemented on an actual motor controller
- None of these classes have been directly tested on physical motor controllers.
- The generic `Servo` class is still not implemented.

### CODE REVIEW

File: Components/Motors/CytronMD30C.cs [hunk 1]
```diff
@@ -2,67 +2,108 @@
...
+        /// Immediately sets the enabled status of the motor. 
+        /// If false, resets TargetSpeed to 0 and stops the motor.
+        /// </summary>
+        public void SetEnabled(bool Enabled)
+        {
+            this.Stopped = !Enabled;
+            if (!Enabled)
+            {
+                this.TargetSpeed = 0;
+                this.SetSpeed(0);
```
Comment (CaiB) [inline Components/Motors/CytronMD30C.cs]: Shouldn't this be `this.SetSpeedDirectly(0);`?

File: Components/Motors/TalonMC.cs [hunk 2]
```diff
@@ -2,62 +2,98 @@
...
 
+        /// <summary> Initializes a Talon Motor controller </summary>
+        /// <param name="PWMOut"> PWM output to control the motor controller </param>
+        /// <param name="MaxSpeed"> Limiting factor for speed (should never exceed + or - this val) </param>
+        /// <param name="SpeedFilter"> Filter to use with MC. Good for ramp-up protection and other applications </param>
         public TalonMC(IPWMOutput PWMOut, float MaxSpeed, IFilter<float> SpeedFilter = null)
         {
             this.PWMOut = PWMOut;
-            this.MaxSpeed = MaxSpeed;
+            this.MaxSpeed = Math.Abs(MaxSpeed);
```
Comment (CaiB) [inline Components/Motors/TalonMC.cs]: Good addition. Missed that they might give us a negative.

File: Filters/Kalman.cs [hunk 3]
```diff
@@ -148,6 +148,13 @@ public void Feed(T Input, T Rate)
             this.Output = SetAngle;
         }
 
+        /// <summary>
+        /// Returns false due to the complex nature of the Kalman filter, 
+        /// since a steady state is difficult to compute.
+        /// </summary>
+        /// <returns> False due to Kalman filter complexity </returns>
```
Comment (CaiB) [inline Filters/Kalman.cs]: Kinda redundant :P

File: Filters/LowPass.cs [hunk 4]
```diff
@@ -75,6 +78,15 @@ public void Feed(T Input, T Rate)
         /// <summary> Resets the low pass filter to the default value of <c>T</c> </summary>
         public void Reset() { this.LastValue = default(T); }
 
+        /// <summary> Computes whether or not the low pass filter is in steady state</summary>
+        /// <returns> Returns whether or not the filter is in steady state</returns>
```
Comment (CaiB) [inline Filters/LowPass.cs]: Space before </tag> please (x2)

File: UnitTest/MotorTesting/TestablePWMOutput.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+        public void Dispose() { }
+
+        public void SetEnabled(bool Enable) { this.Enabled = Enable; }
+
+        public void SetFrequency(int Frequency) { this.Frequency = Frequency; }
+
+        public void SetOutput(float DutyCycle) { this.DutyCycle = DutyCycle; }
+    }
+}
```
Comment (CaiB) [inline UnitTest/MotorTesting/TestablePWMOutput.cs]: Huh, this is kinda like what I had in mind when talking about "virtual I/O" when first implementing this system. If we wanted to later on, we could have I2C buses return data like a sensor would for testing as well.

File: Utilities/UtilConstants.cs [hunk 6]
```diff
@@ -6,6 +6,17 @@
 
 namespace Scarlet.Utilities
 {
+    /// <summary> Class to store constants </summary>
+    public class Constants
+    {
+        // PHYSICAL CONSTANTS
+        public const float GRAVITY = 9.80655f; // meters per second per second
+        public const float EARTH_RADIUS = 6371000; // meters
```
Comment (CaiB) [inline Utilities/UtilConstants.cs]: These aren't used, are they? Kinda random.

File: Filters/LowPass.cs (hunk 4 above)
Comment (CaiB) [reply Filters/LowPass.cs]: Heh, GitHub strips HTML. I meant Space before </tag(greater than)

File: Utilities/UtilConstants.cs (hunk 6 above)
Comment (Baldstrom) [reply Utilities/UtilConstants.cs]: They are used for some other software teams. E.G. GPS coordinate route mapping I believe. I wanted to put in that other constant, but didn't want to make a class for just that. Also if we implement an accelerometer for example, having that gravitational constant would be useful if we wanted to do a conversion before it goes to the user.

Comment (CaiB) [reply Utilities/UtilConstants.cs]: Ah, OK. That makes sense. Document 'em! :P

File: Components/Motors/CytronMD30C.cs (hunk 1 above)
Comment (Baldstrom) [reply Components/Motors/CytronMD30C.cs]: Yes, I will fix that.

**Pull request merged**